### DISCUSSION TITLE
How to call a global JS method?

### DISCUSSION THREAD

**Discussion by faveoled** [Vertispan/j2clmavenplugin#228] [Q&A]
Trying to write a binding to a global JS method with this code:
```java
@JsType(isNative = true, namespace = GLOBAL)
public class Globals {
    @JsMethod(name = "addEventListener", namespace = GLOBAL)
    public static native void addEventListener(String type, EventListener listener);

}
```
JS exports:
```js
var addEventListener = function (type, listener) {};
```
Getting this error:
```
[ERROR] app:app:1.0-SNAPSHOT/optimized_js: /home/user/Dev/Java/j2cl-proxy-workers/target/gwt3BuildCache/0.20/app-app-1.0-SNAPSHOT/fdf0ed43cf36cb92a1a3baa3a34e7934-transpiled_js/results/app/Globals$$Overlay.impl.java.js:13:29: ERROR - [JSC_UNDEFINED_VARIABLE] variable Globals is undeclared
  return instance instanceof Globals;
                             ^^^^^^^

```
Any ideas on how to fix the binding?

**Comment by niloc132** [comment]
The issue here is the same as in https://github.com/Vertispan/j2clmavenplugin/discussions/227 - you're code says "there's a class in JS called `Global`", but then didn't define any externs for that class. Does your project have such a class? Or are you trying to call something like `self.addEventListener` (also could be on `window`, `top`, etc)? If the latter, try Elemental2's `DomGlobal.self.addEventListener()`, which already is defined correctly.

**Reply by faveoled** [reply]
"there's a class in JS called `Global`" is wrong for me, I need to say "there's a global function in JS called `addEventListener`. I'm not sure how to declare it in Java properly. The function should be called as a global function (no `window`, `top` etc.). `DomGlobal` generates this code: `(O(),P).addEventListener` which doesn't look like a global call to me.

**Reply by niloc132** [reply]
If you're running "global" calls in a browser, they exist on "window" or "top" (same thing if you aren't using frames/iframes, but different if you are). If in a web worker or server side, then "self" exists, and globals are usually defined on that. In a browser page, "self" will also point at "window" and "this" (but usually "this" has other meaning in your own method calls, so isn't appropriate).

https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener
https://developer.mozilla.org/en-US/docs/Web/API/Window#instance_methods (see addEventListener)

So, `DomGlobal.self` is probably what you actually want (and if I'm mistaken, can you link to docs showing this?) If this happens to be a non-standard hoisted function, you can cheat by using `@JsMethod(namespace="<window>")` on your `static native` function, but almost certainly this isn't what you actually need.

**Reply by faveoled** [reply]
I believe it's a non-standard function.
https://developers.cloudflare.com/workers/runtime-apis/add-event-listener/
`DomGlobal.self.addEventListener` fails for me with undefined error. 
I've tried `<window` this way:
```java
public class Globals {

    @JsMethod(namespace="<window>", name = "addEventListener")
    public static native void addEventListener(String type, EventListener listener);

}
```
But for some reason it stops generating my method call in JS completely.

**Reply by faveoled** [reply]
One other example I'd like to use a global call in is a fetch Node.js API

**Reply by niloc132** [reply]
The `fetch` api in browsers is the same as in node i think? That is also deceptively also actually part of window/self.

Can you confirm that self.addEventListener gets an "self.addEventListener is not a function" error when run in your cf worker env?

Also, when you say it stops generating your method call, what do you mean? Can you share an example of this? As a native function, declared in the same way as `jsinterop.base.Js#typeof`, it should be emitted correctly.

**Reply by faveoled** [reply]
> The fetch api in browsers is the same as in node i think? That is also deceptively also actually part of window/self.

Probably, I'd have to experiment more

> Can you confirm that self.addEventListener gets an "self.addEventListener is not a function" error when run in your cf worker env?
This code:
```java
DomGlobal.self.addEventListener("fetch", this::handleEvent)
```
Results in this:
```
service core:user:worker: Uncaught TypeError: Cannot read properties of undefined (reading 'addEventListener')
  at core:user:worker:485:18
  at core:user:worker:508:10 in target/app-1.0-SNAPSHOT/app/app.js
  at core:user:worker:7:52 in __require
  at core:user:worker:511:3
  at core:user:worker:512:3

```
> Also, when you say it stops generating your method call, what do you mean? Can you share an example of this? As a native function, declared in the same way as jsinterop.base.Js#typeof, it should be emitted correctly.

I just get this as my target JS file:
```javascript
(function(){/*

 Copyright The Closure Library Authors.
 SPDX-License-Identifier: Apache-2.0
*/
}).call(this);
```
The binding seems similar to `jsinterop.base.Js#typeof`

**Reply by niloc132** [reply]
Indeed it seems as though self doesn't exist. `Js.global()` might give you a real "global" reference, can you see if that also returns undefined?

For your `<window>` experiment, closure seems to have optimized out the function. Were there any warnings in the build? is it possible that it just needs explicit externs? Remember, if you are expecting to use API calls that closure doesn't know about, it must have externs, and you are responsible for defining those (or finding someone else who are already defined them).

**Reply by faveoled** [reply]
`DomGlobal.self.fetch("https://google.com")` fails on Node.js with:
```
TypeError: Cannot read properties of undefined (reading 'fetch')
    at /home/user/Dev/Java/j2cl-node/target/app-1.0-SNAPSHOT/app/app.js:16:23
    at Object.<anonymous> (/home/user/Dev/Java/j2cl-node/target/app-1.0-SNAPSHOT/app/app.js:17:182)
    at Object.<anonymous> (/home/user/Dev/Java/j2cl-node/target/app-1.0-SNAPSHOT/app/app.js:17:332)
    at Module._compile (node:internal/modules/cjs/loader:1159:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1213:10)
    at Module.load (node:internal/modules/cjs/loader:1037:32)
    at Module._load (node:internal/modules/cjs/loader:878:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
    at node:internal/main/run_main_module:23:47
```

**Reply by faveoled** [reply]
> Js.global() might give you a real "global" reference, can you see if that also returns undefined?

`Js.global()` is an object

> Were there any warnings in the build? is it possible that it just needs explicit externs? Remember, if you are expecting to use API calls that closure doesn't know about, it must have externs, and you are responsible for defining those (or finding someone else who are already defined them).

No warnings, there's an extern but I'm not knowledgeable enough to know if it's picked up

**Reply by niloc132** [reply]
Note that this is the same error, `self` isn't defined. There might be some other global reference you can use in its place, but I don't write enough server-side JS to know what it would be at this time. Poking quickly at a node console, I see `this` is defined, and has `fetch()` defined on it, and goog.global (aka Js.global()) would access it to..

I do see `global` defined though, and it is a reference to `this`, so `global.fetch()` should work. That is, make a public static field like this:

```
@JsType(isNative = true, name = "global", namespace = JsPackage.GLOBAL)
public class NodeGlobal {
  public static Window global;
}
```

That would technically let you reference js `global.global` (which exists), and then use other functions on window.

With that said.... DomGlobal.fetch() is a function, and it is declared as actually on the global namespace (via goog.global), so I think that will work as-is - you don't need DomGlobal.self.fetch().

**Reply by faveoled** [reply]
Still haven't got any luck (will try again sometimes in the future)
```java
DomGlobal.fetch("https://google.com")
```
```
TypeError: ha.fetch is not a function
    at /home/user/Dev/Java/j2cl-node/target/app-1.0-SNAPSHOT/app/app.js:16:16
    at Object.<anonymous> (/home/user/Dev/Java/j2cl-node/target/app-1.0-SNAPSHOT/app/app.js:17:159)
    at Object.<anonymous> (/home/user/Dev/Java/j2cl-node/target/app-1.0-SNAPSHOT/app/app.js:17:171)
    at Module._compile (node:internal/modules/cjs/loader:1159:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1213:10)
    at Module.load (node:internal/modules/cjs/loader:1037:32)
    at Module._load (node:internal/modules/cjs/loader:878:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
    at node:internal/main/run_main_module:23:47
```

```java
NodeGlobal.global.fetch("https://google.com")
```
```
TypeError: Cannot read properties of undefined (reading 'fetch')
    at /home/user/Dev/Java/j2cl-node/target/app-1.0-SNAPSHOT/app/app.js:16:23
    at Object.<anonymous> (/home/user/Dev/Java/j2cl-node/target/app-1.0-SNAPSHOT/app/app.js:17:182)
    at Object.<anonymous> (/home/user/Dev/Java/j2cl-node/target/app-1.0-SNAPSHOT/app/app.js:17:194)
    at Module._compile (node:internal/modules/cjs/loader:1159:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1213:10)
    at Module.load (node:internal/modules/cjs/loader:1037:32)
    at Module._load (node:internal/modules/cjs/loader:878:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
    at node:internal/main/run_main_module:23:47
```

**Reply by niloc132** [reply]
Did the externs solve your global addEventListener problem at least? I didnt see you respond to that part, or to the question about warnings in the build.

**Reply by faveoled** [reply]
Externs help with a compilation failure, but I haven't succeeded with global JS calls. In the case of empty JS output no warnings were there

**Reply by niloc132** [reply]
Okay - sounds like we could do with a test case to confirm if our repo or forks introduced something, if there is an upstream bug, or something subtle wrong with your code.
